homeassistant/components/nws/__init__.py

```diff
@@ -52,38 +52,54 @@ def signal_unique_id(latitude, longitude):
 
 
 async def async_setup(hass: HomeAssistant, config: dict):
-    """Set up the National Weather Service integration."""
-    if DOMAIN not in config:
-        return True
-
-    hass.data[DOMAIN] = hass.data.get(DOMAIN, {})
-    for entry in config[DOMAIN]:
-        latitude = entry.get(CONF_LATITUDE, hass.config.latitude)
-        longitude = entry.get(CONF_LONGITUDE, hass.config.longitude)
-        api_key = entry[CONF_API_KEY]
-
-        client_session = async_get_clientsession(hass)
-
-        if base_unique_id(latitude, longitude) in hass.data[DOMAIN]:
-            _LOGGER.error(
-                "Duplicate entry in config: latitude %s  latitude: %s",
-                latitude,
-                longitude,
-            )
-            continue
-
-        nws_data = NwsData(hass, latitude, longitude, api_key, client_session)
-        hass.data[DOMAIN][base_unique_id(latitude, longitude)] = nws_data
-        async_track_time_interval(hass, nws_data.async_update, DEFAULT_SCAN_INTERVAL)
-
-        for component in PLATFORMS:
-            hass.async_create_task(
-                discovery.async_load_platform(hass, component, DOMAIN, entry, config)
-            )
+    """Set up the National Weather Service (NWS) component."""
+    hass.data.setdefault(DOMAIN, {})
+    return True
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry):
+    """Set up a National Weather Service entry."""
+    latitude = entry.data[CONF_LATITUDE]
+    longitude = entry.data[CONF_LONGITUDE]
+    api_key = entry.data[CONF_API_KEY]
+    station = entry.data[CONF_STATION]
+
+    client_session = async_get_clientsession(hass)
+
+    nws_data = NwsData(hass, latitude, longitude, api_key, client_session)
+    hass.data[DOMAIN][entry.entry_id] = nws_data
+
+    # async_set_station only does IO when station is None
+    await nws_data.async_set_station(station)
+    await nws_data.async_update()
```

Comment:
I've been noticing that the nws servers can be a bit flakey when things are busy so the config entry isn't going to get setup and won't be retried if nws is having issues

```
2020-04-15 18:26:44 INFO (MainThread) [homeassistant.setup] Setting up nws
2020-04-15 18:26:44 INFO (MainThread) [homeassistant.setup] Setup of domain nws took 0.0 seconds.
2020-04-15 18:26:49 INFO (MainThread) [homeassistant.components.weather] Setting up weather.nws
2020-04-15 19:27:07 WARNING (MainThread) [homeassistant.components.nws] Error updating observation for station KHOU: 502, message='Bad Gateway', url='https://api.weather.gov/stations/KHOU/observations
2020-04-15 19:36:51 WARNING (MainThread) [homeassistant.components.nws] Success updating observation for station KHOU
```

I'm not sure `asyncio.Timeout` is the only exception you need to check though.

```suggestion
    try:
        await nws_data.async_update()
    except asyncio.Timeout:
        raise ConfigEntryNotReady
```

Reply:
I think that this update failing won't prevent setup, it will just prevent data being available upon startup.  The update failing that would prevent setup is done inside validate_input in config_flow.  The proof is in the tests inside test_weather (these tests probably make more sense to be in test_init) where the first call to update the data fails and everything sets up OK.

Reply:
It works much better in the config flow framework since the only thing that prevents setup is if the station cannot be retrieved, but I think in the config flow framework, the station is saved after configuration and never needs to be done again?

Reply:
If `await nws_data.async_update()` can throw an exception it should be wrapped with the try/except here.  If its can't ever throw, no need to do it.

Reply:
You'll get something like this in the log and home assistant will try again in X seconds if you throw `ConfigEntryNotReady` 
`2020-04-15 08:32:46 WARNING (MainThread) [homeassistant.config_entries] Config entry for nws not ready yet. Retrying in 10 seconds.`

Reply:
The IO is wrapped here already:  do you have other exceptions we should catch?  The ClientError seems to be pretty broad.

https://github.com/home-assistant/core/pull/34267/files#diff-53e72c9a830e23639d1a7bd81cd66380L136-L152

Reply:
It seems you cannot point directly to a particular line in a diff in a PR, or I cannot figure it out, it is currently here 
https://github.com/home-assistant/core/blob/dev/homeassistant/components/nws/__init__.py#L136-L152

Reply:
I see you are already trapping them.  Let me update the suggestion

Reply:
As I understand it, the current behavior should set up if an error occurs but no data would be available.  Why isn't this the preferred behavior?

Reply:
The entity will be set to Unavailable if you raise `ConfigEntryNotReady`. If you don't raise it, it will be set to Unknown which might impact user's automations.   I'm not sure it makes too much difference in this case since its more of a problem with integrations that don't have enough data to setup entities.

https://developers.home-assistant.io/docs/integration_quality_scale_index/

Reply:
I think the entities will already report `unavailable` as the `available` property will return `False` .  Let me see if I can add this to the tests to check.

Reply:
I added tests to confirm my understanding.  If there is a still a reason to raise the error, I can easily add it, but it seems unnecessary at the moment.

Reply:
This isn't a case where the entities will be broken because the update fails so that works for me.